usr/sbin/pkcsep11_session/pkcsep11_session.c

```diff
@@ -90,120 +90,61 @@ CK_VERSION lib_version;
 #define ACTION_VHSMPIN  3
 #define ACTION_STATUS   4
 
-int get_pin(char **pin, size_t *pinlen)
-{
-    struct termios old, new;
-    int nread;
-    char *buff = NULL;
-    size_t buflen;
-    int rc = 0;
-
-    /* turn echoing off */
-    if (tcgetattr(fileno(stdin), &old) != 0)
-        return -1;
-
-    new = old;
-    new.c_lflag &= ~ECHO;
-    if (tcsetattr(fileno(stdin), TCSAFLUSH, &new) != 0)
-        return -1;
-
-    /* read the pin
-     * Note: getline will allocate memory for buff. free it when done.
-     */
-    nread = getline(&buff, &buflen, stdin);
-    if (nread == -1) {
-        rc = -1;
-        goto done;
-    }
-
-    /* Restore terminal */
-    tcsetattr(fileno(stdin), TCSAFLUSH, &old);
-
-    /* start a newline */
-    printf("\n");
-    fflush(stdout);
-
-    /* Allocate  PIN.
-     * Note: nread includes carriage return.
-     * Replace with terminating NULL.
-     */
-    *pin = (char *) malloc(nread);
-    if (*pin == NULL) {
-        rc = -ENOMEM;
-        goto done;
-    }
-
-    /* strip the carriage return since not part of pin. */
-    buff[nread - 1] = '\0';
-    memcpy(*pin, buff, nread);
-    /* don't include the terminating null in the pinlen */
-    *pinlen = nread - 1;
-
-done:
-    if (buff)
-        free(buff);
-
-    return rc;
-}
-
 static int get_user_pin(CK_BYTE *dest)
 {
-    int ret;
-    char *userpin = NULL;
-    size_t userpinlen;
-
-    printf("Enter the USER PIN: ");
-    fflush(stdout);
-    ret = get_pin(&userpin, &userpinlen);
-    if (ret != 0) {
+    int ret = -1;
+    const char *userpin;
+    struct pinbuf buf_user = pinbuf_empty;
+
+    userpin = pin_get_user(&buf_user);
+    if (!userpin) {
         fprintf(stderr, "Could not get USER PIN.\n");
-        return -1;
+        goto out;
     }
 
-    if (userpinlen > PKCS11_MAX_PIN_LEN) {
+    if (strlen(userpin) > PKCS11_MAX_PIN_LEN) {
         fprintf(stderr, "The USER PIN must be less than %d chars in length.\n",
                 (int) PKCS11_MAX_PIN_LEN);
-        free(userpin);
-        return -1;
+        goto out;
     }
 
-    memcpy(dest, userpin, userpinlen + 1);
-    free(userpin);
-
-    return 0;
+    memcpy(dest, userpin, strlen(userpin) + 1);
+    ret = 0;
+out:
+    pinbuf_free(&buf_user);
+    return ret;
 }
 
 static int get_vhsm_pin(CK_BYTE *dest)
 {
-    int ret;
-    char *vhsmpin = NULL;
+    int ret = -1;
+    const char *vhsmpin = NULL;
+    struct pinbuf buf_vhsm = pinbuf_empty;
     size_t vhsmpinlen;
 
-    printf("Enter the new VHSM PIN: ");
-    fflush(stdout);
-    ret = get_pin(&vhsmpin, &vhsmpinlen);
-    if (ret != 0) {
+    vhsmpin = pin_prompt(&buf_vhsm, "Enter the new VHSM PIN: ");
```

Comment:
I guess we should be using pin_get_new() here.... (although it was not that way before, I know). We are asking the user to set a new pin, so maybe its better to require a confirmation of the pin entered here.

Reply:
Will need an update in testcases/init_vhsm.exp.in as well to provide the pin again to the 2nd prompt.

Reply:
ok, I will have a look at that. Maybe I'll come back to you for the details.

Reply:
This will not be in the next version, but I keep it on my to-do list.